load.php

```diff
@@ -226,7 +226,7 @@ function perflab_render_generator() {
  * @since 1.3.0
  *
  * @param string $module Slug of the module.
- * @return bool Whether the module can be loaded or not.
+ * @return bool|WP_Error Whether the module can be loaded or not, otherwise a WP_Error.
```

Comment:
I personally don't like the idea of overloading this function to return an error if the module is already loaded via a standalone plugin. Instead, this function should return false if the module code is already loaded.

Here's what I would propose:

1. Create a new function, like `perflab_is_module_standalone( string $slug ): bool {}` that checks whether the module slug's constant is already defined.
2. Use this function inside `perflab_can_load_module()` to decide whether to return `false` in place of the current logic that throws a WP_Error.
3. Replace the else condition in `perflab_render_modules_page_field` that looks for WP_Errors with a new condition that checks for a false response from `perflab_is_module_standalone` and displays a message that says the module can't be activated because the code is being loaded by a separate plugin.

Reply:
Another concern I have is that we are also using `perflab_can_load_module()` in `perflab_is_valid_module()` to determine if we should display the module as generator content in `perflab_get_generator_content()`. Wouldn't we still want to show that a feature is active on a site regardless of whether it's loaded as a Perf Lab module or a standalone plugin?

Reply:
I'm okay not returning a `WP_Error`, but then I also think we shouldn't intertwine the two functions at all. I think a simple boolean lacks the necessary context we would want to provide in end-user facing messaging, which is why I suggested the `WP_Error` usage here.

Alternatively, we can always use only `bool`s, but then it should be individual functions that are individually called, and I don't think we should put the `perflab_is_module_standalone()` call into `perflab_can_load_module()` as otherwise the context is again intertwined and the return value does not convey _why_.

While the above might be okay, I still think even with that it has drawbacks compared to using a `WP_Error`: When you look at our `admin/load.php` file, you see all those manual messages in there for SQLite, which with a `WP_Error` return value we could work towards avoiding those. Those messages should rather be located where those problems are determined, not manually be re-checked when we already have that logic somewhere else. I think using `false` instead of `WP_Error` lacks context and will require us to repeat code unnecessarily. What is the concern about returning a `WP_Error`?

Reply:
> What is the concern about returning a WP_Error?

I have two main concerns:

1. It complicates the function signature and purpose unnecessarily and requires any code using this function to check for a possible error message OR a boolean value, when the intent of the function is mainly to determine if a module can be loaded. Keeping the function simple makes it easier to reason about and easier to test to avoid future bugs.
2. `WP_Error` messages should be used when we actually _want_ to relay that an error has occurred. In this case, we're using it as a logical value in our code so that we can _avoid_ an error. That seems wrong.

I'd be happy to decouple all of this logic as well, if you'd prefer, I just think it would be weird for a function called `perflab_can_load_module()` to return `true` if we know that the code has already been loaded, which is why I'm proposing the approach I suggested.

Reply:
Thanks @joemcgill and @felixarntz for the feedback. It look promising to me and I implemented it in https://github.com/WordPress/performance/pull/670/commits/0c5bc48c15ddf9703dfc5fb6880bedb9d577d0e2. It removes the `WP_Error` and introduce new function `perflab_is_module_standalone` as suggested.

Can we open a new issue and work on SQLite module stuff if it isn't a blocker for this issue? 

Reply:
Thanks @mukeshpanchal27, a new follow up issue sounds like a great idea. We can use that issue to implement a better architecture for handling _why_ a module cannot be loaded and making this information available to the end user.

The approach you implemented looks good, I just left a review in https://github.com/WordPress/performance/pull/670#pullrequestreview-1348814786.